### PR TITLE
[dependencies] Drop support for python2 #140

### LINKED ISSUE #140: [dependencies] Drop support for python 2
Set travis to test on python 3.7.
Drop support for python 2 and its constructs.
Simplify `super()` calls.

### PR DESCRIPTION
**Pull request by IAmAMissingColon**
* Simplified super() function
* removed 2.7 from travis
* added 3.7 in travis

Fixes #140

### CODE REVIEW

File: netjsonconfig/backends/openwrt/converters/openvpn.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ def __intermediate_vpn(self, vpn):
             '.type': 'openvpn',
             'enabled': not vpn.pop('disabled', False)
         })
-        return super(OpenVpn, self).__intermediate_vpn(vpn, remove=[''])
```
Comment (nemesisdesign) [inline netjsonconfig/backends/openwrt/converters/openvpn.py]: I see another occurrence of `super` in this file at line 23

File: .travis.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@ cache: pip
 python:
   - "3.5"
   - "3.4"
```
Comment (nemesisdesign) [inline .travis.yml]: add `3.6`, remove `3.4` and `3.5`.

File: netjsonconfig/backends/openwrt/converters/openvpn.py (hunk 1 above)
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/openvpn.py]: Done :smile:

File: .travis.yml (hunk 2 above)
Comment (IAmAMissingColon) [reply .travis.yml]: :+1: +

File: setup.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 if sys.argv[-1] == 'publish':
     if sys.version_info.major < 3:
         print('Error: do not publish release with python2!')
-        sys.exit(1)
+        sys.exit(1)
```
Comment (nemesisdesign) [inline setup.py]: I think there's a few extra spaces at the end of this line which shouldn't be here

File: setup.py [hunk 4]
```diff
@@ -13,9 +13,6 @@
     print("To install, run 'python setup.py install'\n")
 
 if sys.argv[-1] == 'publish':
-    if sys.version_info.major < 3:
```
Comment (IAmAMissingColon) [inline setup.py]: I'm confused on this we need to remove this right ? because we are dropping support of python2

**Comment by IAmAMissingColon** [comment]
I see `python2` and `python 2` but they are in `#`

**Comment by IAmAMissingColon** [comment]
should I remove them as well ?

**Comment by IAmAMissingColon** [comment]
`ERROR: coveralls 1.9.2 has requirement coverage<5.0,>=3.6, but you'll have coverage 5.0 which is incompatible.`

fixing this :smile:

**Comment by IAmAMissingColon** [comment]
> Look for `python2:` as well.

the serach gives output something like this :+1: 

```
Searching 111 files for "python2:"

/home/runtimeterror/Desktop/task_7/netjsonconfig/bin/netjsonconfig:
  147      if hasattr(output, 'getvalue'):
  148          output = output.getvalue()
  149:     # python2: always print
  150      # python3: if string, just print
  151      if isinstance(output, six.string_types):
  152          print(output)
  153:     # python2: never enters this block
  154      # python3: writes binary data to standard output
  155      else:

2 matches in 1 file

```

File: setup.py [hunk 5]
```diff
@@ -78,7 +66,6 @@ def get_install_requires():
         'Topic :: System :: Networking',
     ],
     install_requires=get_install_requires(),
-    extras_require=extras_require,
     test_suite='nose2.collector.collector',
     scripts=['bin/netjsonconfig'],
-)
+)
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply setup.py]: Sorry !

**Comment by IAmAMissingColon** [comment]
> > Look for `python2:` as well.
> 
> the serach gives output something like this:
> 
> ```
> Searching 111 files for "python2:"
> 
> /home/runtimeterror/Desktop/task_7/netjsonconfig/bin/netjsonconfig:
>   147      if hasattr(output, 'getvalue'):
>   148          output = output.getvalue()
>   149:     # python2: always print
>   150      # python3: if string, just print
>   151      if isinstance(output, six.string_types):
>   152          print(output)
>   153:     # python2: never enters this block
>   154      # python3: writes binary data to standard output
>   155      else:
> 
> 2 matches in 1 file
> ```

As I think the code is for both versions, but the comments are just indicating what the code will do in python2 and python3 

please correct me if I'm wrong.. It will improve me :smile:

**Comment by atb00ker** [comment]
> As I think the code is for both versions, but the comments are just indicating what the code will do in python2 and python3

1. Read about python six. Do we need it?
2. Read the function, see what it's doing for python2. You are dropping python 2 so do you need the code for python 2 now?

File: setup.py [hunk 6]
```diff
@@ -78,7 +66,7 @@ def get_install_requires():
         'Topic :: System :: Networking',
     ],
     install_requires=get_install_requires(),
-    extras_require=extras_require,
     test_suite='nose2.collector.collector',
     scripts=['bin/netjsonconfig'],
-)
+
```
Comment (nemesisdesign) [inline setup.py]: avoid adding the blank line

File: setup.py [hunk 7]
```diff
@@ -78,7 +66,7 @@ def get_install_requires():
...
+)
```
Comment (nemesisdesign) [inline setup.py]: ensure the file ends with a new line and configure your editor to always add new lines at the end of files

Comment (IAmAMissingColon) [reply setup.py]: Okay :smile:

File: setup.py (hunk 6 above)
Comment (IAmAMissingColon) [reply setup.py]: Done :smile:

**Comment by IAmAMissingColon** [comment]
> @IAmAMissingColon some tests are failing, can you find out the cause and fix them please?

Done :smile:

**Comment by IAmAMissingColon** [comment]
Build was failing because of removal of six

File: bin/netjsonconfig [hunk 8]
```diff
@@ -146,11 +146,9 @@ def print_output(output):
     # if file object, get bytes
     if hasattr(output, 'getvalue'):
         output = output.getvalue()
-    # python2: always print
     # python3: if string, just print
     if isinstance(output, six.string_types):
```
Comment (atb00ker) [inline bin/netjsonconfig]: I don't know the source of the confustion but, I have given you multiple hints about six.
There is the direct answer that I don't like giving.
"Six is a Python 2 and 3 compatibility library. It provides utility functions for smoothing over the differences between the Python versions with the goal of writing Python code that is compatible on both Python versions. "
- Source: Your submission on GCI, https://pypi.org/project/six/1.13.0/

How I am have different research task for you, do we need python2 compatibility if we are dropping python 2?

**Comment by IAmAMissingColon** [comment]
Replaced `six.string_types` with `str`

File: bin/netjsonconfig [hunk 9]
```diff
@@ -146,11 +144,9 @@ def print_output(output):
     # if file object, get bytes
     if hasattr(output, 'getvalue'):
         output = output.getvalue()
-    # python2: always print
     # python3: if string, just print
```
Comment (atb00ker) [inline bin/netjsonconfig]: Comment not required anymore.

File: bin/netjsonconfig [hunk 10]
```diff
@@ -146,11 +144,9 @@ def print_output(output):
...
-    if isinstance(output, six.string_types):
+    if isinstance(output, str):
         print(output)
-    # python2: never enters this block
     # python3: writes binary data to standard output
```
Comment (atb00ker) [inline bin/netjsonconfig]: Comment not required anymore.

Comment (IAmAMissingColon) [reply bin/netjsonconfig]: Hm! the comment is not present in my local change why it's present in here ?

File: netjsonconfig/backends/openwrt/converters/interfaces.py [hunk 11]
```diff
@@ -68,7 +66,7 @@ def __intermediate_addresses(self, interface):
             # do not use CIDR notation when using a single ipv4
             # see https://github.com/openwisp/netjsonconfig/issues/54
             if len(static.get('ipaddr', [])) == 1:
-                network = ip_interface(six.text_type(static['ipaddr'][0]))
+                network = ip_interface(str(static['ipaddr'][0]))
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/interfaces.py]: `type()` shows that it's not a string

File: netjsonconfig/backends/openwrt/converters/interfaces.py [hunk 12]
```diff
@@ -273,7 +271,7 @@ def __netjson_addresses(self, interface):
         return interface
 
     def __netjson_address(self, address, interface):
-        ip = ip_interface(six.text_type(address))
+        ip = ip_interface(str(address))
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/interfaces.py]: type() shows that it's not a string, you are right

File: netjsonconfig/backends/openwrt/converters/rules.py [hunk 13]
```diff
@@ -23,9 +21,9 @@ def __intermediate_rule(self, rule, index):
         dest_net = None
         family = 4
         if 'src' in rule:
-            src_net = ip_network(six.text_type(rule['src']))
+            src_net = ip_network(str(rule['src']))
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/rules.py]: type() shows that it's not a string, you are right

File: netjsonconfig/backends/openwrt/converters/rules.py [hunk 14]
```diff
@@ -23,9 +21,9 @@ def __intermediate_rule(self, rule, index):
...
         if 'dest' in rule:
-            dest_net = ip_network(six.text_type(rule['dest']))
+            dest_net = ip_network(str(rule['dest']))
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/rules.py]: type() shows that it's not a string, you are right

File: netjsonconfig/backends/openwrt/converters/routes.py [hunk 15]
```diff
@@ -52,7 +50,7 @@ def __netjson_route(self, route, i):
             network = '{0}/{1}'.format(network, route.pop('netmask'))
         route.update({
             "device": route.pop('interface'),
-            "destination": str(ip_interface(six.text_type(network))),
+            "destination": str(ip_interface(str(network))),
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/routes.py]: type() shows that it's not a string, you are right

File: netjsonconfig/backends/openwrt/converters/routes.py [hunk 16]
```diff
@@ -18,7 +16,7 @@ def to_intermediate_loop(self, block, result, index=None):
         return result
 
     def __intermediate_route(self, route, index):
-        network = ip_interface(six.text_type(route.pop('destination')))
+        network = ip_interface(str(route.pop('destination')))
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/routes.py]: `type()` shows that it's not a string, you are right

File: netjsonconfig/backends/openwrt/converters/interfaces.py [hunk 17]
```diff
@@ -283,7 +281,7 @@ def __netjson_address(self, address, interface):
         ))
         uci_gateway_key = 'gateway' if family == 'ipv4' else 'ip6gw'
         gateway = interface.get(uci_gateway_key, None)
-        if gateway and ip_address(six.text_type(gateway)) in ip.network:
+        if gateway and ip_address(str(gateway)) in ip.network:
```
[the remark this answers is not in the document]
Comment (IAmAMissingColon) [reply netjsonconfig/backends/openwrt/converters/interfaces.py]: `type()` shows that it's not a string, you are right

File: bin/netjsonconfig (hunk 9 above)
Comment (IAmAMissingColon) [reply bin/netjsonconfig]: The comment is not present in my local repo, I remved it and pushed it before ! Hmm! but why it's showing in here ?

Comment (atb00ker) [reply bin/netjsonconfig]: I am talking about "# python3: if string, just print"; It's there, I can guarantee you.
You might have looked into some other file by mistake!
Please check again. Send screenshot if you don't see it! :smile:

File: bin/netjsonconfig (hunk 10 above)
Comment (atb00ker) [reply bin/netjsonconfig]: I am talking about "# python3: writes binary data to standard output".

**Comment by IAmAMissingColon** [comment]
Thanks :smile:

**Pull request merged**